src/ripple/nodestore/ShardInfo.h

```diff
@@ -0,0 +1,67 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_SHARDINFO_H_INCLUDED
+#define RIPPLE_NODESTORE_SHARDINFO_H_INCLUDED
+
+#include <ripple/basics/RangeSet.h>
+#include <ripple/nodestore/Types.h>
+#include <ripple/nodestore/impl/Shard.h>
+#include <ripple/protocol/messages.h>
+
+namespace ripple {
+namespace NodeStore {
+
+/* Contains information on the status of shards for a node
+ */
+class ShardInfo
```

Comment:
While reviewing the code I found a couple of different places directly manipulating the members of the `ShardInfo`.  Here's an example of what I saw:
```
            switch (state)
            {
                case final:
                    shardInfo->final.insert(shard->index());
                    break;

                default:
                    shardInfo->incomplete.emplace(
                        shard->index(),
                        ShardInfo::Incomplete(state, shard->getProgress()));
                    break;
            }
```
The data manipulations are subtle enough that I think it may be worthwhile giving `ShardInfo` a member function that handles the update and then making the data members private.  Here's what the update member function ended up looking like for me:
```
    // Returns true if update is successful.  Returns false if update fails
    // because of a duplicate incomplete index.
    bool
    update(
        std::uint32_t index,
        ShardState state,
        Shard::Progress const& progress)
    {
        if (state == ShardState::finalized)
        {
            finalized_.insert(index);
            return true;
        }
        return incomplete_.emplace(index, Incomplete(state, progress)).second;
    }
```
I just ended up feeling more comfortable when this operation was kept in a central place.  Your milage may vary.

After making the data members `private` I needed to provide accessor functions.  But at least those accessors could provide `const` references, so all changes to the member data go through the interface.

I also needed to provide one additional manipulator method, since there's a place in the code that initializes the `finalized_` member from a string:
```
    [[nodiscard]] bool
    setFinalizedFromString(std::string const& str)
    {
        return from_string(finalized_, str);
    }
```
My eventual version for the entire header ended up looking like this:
```
class ShardInfo
{
private:
    class Incomplete
    {
    public:
        Incomplete() = delete;
        Incomplete(ShardState state_, Shard::Progress const& progress)
            : state(state_)
            , percentProgress(
                  progress.target > 0u
                      ? std::min(99u, (progress.soFar * 100u) / progress.target)
                      : 0u)
        {
        }

        ShardState const state;
        std::uint32_t const percentProgress;
    };

public:
    [[nodiscard]] RangeSet<std::uint32_t> const&
    finalized() const
    {
        return finalized_;
    }

    [[nodiscard]] std::map<std::uint32_t, Incomplete> const&
    incomplete() const
    {
        return incomplete_;
    }

    [[nodiscard]] std::string
    finalizedToString() const;

    [[nodiscard]] std::string
    incompleteToString() const;

    [[nodiscard]] bool
    setFinalizedFromString(std::string const& str)
    {
        return from_string(finalized_, str);
    }

    [[nodiscard]] protocol::TMPeerShardInfoV2
    makeMessage(Application& app) const;

    // Returns true if update is successful.  Returns false if update fails
    // because of a duplicate incomplete index.
    bool
    update(
        std::uint32_t index,
        ShardState state,
        Shard::Progress const& progress)
    {
        if (state == ShardState::finalized)
        {
            finalized_.insert(index);
            return true;
        }
        return incomplete_.emplace(index, Incomplete(state, progress)).second;
    }

private:
    // Complete and verified immutable shards
    RangeSet<std::uint32_t> finalized_;

    // Incomplete shards being acquired or verified
    std::map<std::uint32_t, Incomplete> incomplete_;
};
```
Your call about whether you want to adopt these changes, but I thought you should at least see what I came up with.